### ISSUE TITLE
stop ignoring how many times moves appear in the data

### ISSUE THREAD

**Issue by Beno71** [stefanpeidli/GoNet#18]
Until now we treated all entries in our boar-move-dictionary in the same manner. We need to take into account that different boards occur differently often in the game data.

example: empty board appears in every game whereas boards with 50+ stones already played occure only in a few of the games.

possible ways of handling the problem:

1.  learn moves n times in a row where n is the number of games they appear in. **problem:** Won't do because with using batches our NN would be trained too strong in one direction, i.e. would forget things it learned before.
2. learn moves in random order where the probability of a move being chosen depends on its total occurences in the game data. **problem:** How should we implement this?

**Comment by stefanpeidli** [comment]
Let's put a little more structure on that:

What's our problem?:
Assume we have 100 different boards in a TrainingSet. Let's also assume that the first board has 99 moves played on it and every other only one move. We would want the Neuro Net now to be **very** good at the first board, as it appears as often as all other boards together. So it needs to be learned more thoroughly. How do we do this?

Observations/Suggestions:
1) Learning the first board 99 times in a row without changing weights is equal to learning it only once, but with learningrate * 99. As stated above, this might lead to **poor learning**. Also we might have to scale down the learning rate accordingly. (_approach 1 above_)

2) Learning *moves* instead of boards with moves on them was another idea. We are following this approach with rather limited success. Inherent problem is imo that moves are **not the correct points** in the approximation space.

3) Yet a new idea would be creating a new TrainingSet. We build it like this: The first board appears 99 times and every other board once, yielding a new size of the Set of 198. Then as usual we decompose the Set into batches and learn them randomly. On a heuristical stance this equals learning words that are important more often, because they are important as they appar more frequently than other words. **Caution: This approach will lead to immensely larger Sets**.

4) Following the last idea we might want to **restrict the size of the Sets** to keep them managable to work with. Let's fix a certain limit like 100<**120**<198. Now instead of learning all 198 boards, we construct a new Set by **uniformly and randomly** choosing 120 boards from the 198 ones, and learn those instead. (_apporach 2 above_)

**Comment by stefanpeidli** [comment]
So we have converged into two solutions:
1) We **increase the pool of elements** to be learned by adding multiples of boards that occur often. 
Assume we have only two distinct boards to learn, b1 and b2, occuring once and twice. Then we would create a pool of [b1,b2,b2] which is to be learned. By shuffling this pool before we assign batches, we ensure that important boards are very frequently (but spreaded across the learning process) been learned. This could be easily done with **DBs**. 

2) We learn boards that are important (=played on often) with a bigger, **adaptive step size** for the gradient descent. This forces the Net to learn them more thoroughly. Also one could change the error measurement accordingly to give a better fit. This might be the simplest solution, but increasing the step size **potentially decreases the efficience of the gradient descent method**. Step size rules could be for example a linear or a logarithmic function.

Testing will show which method is best.

**Comment by Beno71** [comment]
issue can be closed, we are done regarding data structure
